core/Types.h

```diff
@@ -707,8 +719,9 @@ TYPE(OrType) final : public Refcounted {
     friend TypePtr Types::Boolean();
     friend class NameSubstitution;
     friend class serialize::SerializerImpl;
+    template <class T>
     friend bool Types::isSubTypeUnderConstraint(const GlobalState &gs, TypeConstraint &constr, const TypePtr &t1,
-                                                const TypePtr &t2, UntypedMode mode);
+                                                const TypePtr &t2, UntypedMode mode, T &errorCollector);
```

Comment:
One of them is instantiated here: https://github.com/sorbet/sorbet/pull/7620/files#diff-eb2da3e78eea17626fb563370b444cc444ebc51aa7b8cd35da2fad4e43cbdd36R1517
The other is (implicitly) instantiated here: https://github.com/sorbet/sorbet/pull/7620/files#diff-eb2da3e78eea17626fb563370b444cc444ebc51aa7b8cd35da2fad4e43cbdd36R1386

Reply:
It's fine to have the function declared without an implementation: whatever types we call that function with, the compiler will turn into a call to a named function that must exist at link time.  If it doesn't exist at link time, of course, you'll get peculiar errors that might not be the most straightforward to track down.

This situation is a little different, because it only declares that `Types::isSubTypeUnderConstraint` -- with any template parameter -- is permitted to access the private members of `OrType`, which seems appropriate.  This declaration is purely for semantic checks and has no effect on code generation.  It's...not the best, as we're now leaking what are supposed to be internal implementation details of `isSubTypeUnderConstraint` into the types.  We could just do `friend class Types` and drop the other declarations of `Types` static methods below?